solidity/contracts/bridge/Bridge.sol

```diff
@@ -778,26 +939,618 @@ contract Bridge is Ownable {
+        // referring back to the wallet PKH and just burning main UTXO value
+        // for transaction fees.
+        require(
+            processedRedemptionsCount > 0,
+            "Redemption transaction must process at least one redemption"
+        );
+
+        return info;
     }
 
-    // TODO It is possible a malicious wallet can sweep deposits that can not
-    //      be later proved on Ethereum. For example, a deposit with
-    //      an incorrect amount revealed. We need to provide a function for honest
-    //      depositors, next to sweep, to prove their swept balances on Ethereum
-    //      selectively, based on deposits they have earlier received.
-    //      (UPDATE PR #90: Is it still the case since amounts are inferred?)
+    // TODO: Function `notifyRedemptionTimeout. That function must:
+    //       1. Take a the `walletPubKey` and `redeemerOutputScript` as params.
+    //       2. Build the redemption key using those params.
+    //       3. Use the redemption key and take the request from
+    //          `pendingRedemptions` mapping.
+    //       4. If request doesn't exist in mapping - revert.
+    //       5. If request exits, and is timed out - remove the redemption key
+    //          from `pendingRedemptions` and put it to `timedOutRedemptions`
+    //          by copying the entire `RedemptionRequest` struct there. No need
+    //          to check if `timedOutRedemptions` mapping already contains
+    //          that key because `requestRedemption` blocks requests targeting
+    //          non-active wallets. Because `notifyRedemptionTimeout` changes
+    //          wallet state after first call (point 9), there is no possibility
+    //          that the given redemption key could be reported as timed out
+    //          multiple times. At the same time, if the given redemption key
+    //          was already marked as fraudulent due to an amount-related fraud,
+    //          it will not be possible to report a time out on it since it
+    //          won't be present in `pendingRedemptions` mapping.
+    //       6. Return the `requestedAmount` to the `redeemer`.
+    //       7. Reduce the `pendingRedemptionsValue` (`wallets` mapping) for
+    //          given wallet by request's redeemable amount computed as
+    //          `requestedAmount - treasuryFee`.
+    //       8. Punish the wallet, probably by slashing its operators.
+    //       9. Change wallet's state in `wallets` mapping to `MovingFunds` in
+    //          order to prevent against new redemption requests hitting
+    //          that wallet.
+    //      10. Expect the wallet to transfer its funds to another healthy
+    //          wallet (just as in case of failed heartbeat). The wallet is
+    //          expected to finish the already queued redemption requests
+    //          before moving funds but we are not going to enforce it on-chain.
+
+    // TODO: Function `submitRedemptionFraudProof`. That function must:
```

Comment:
I suggest we leave the following note to auditors in this TODO:

Deposit and redemption fraud proofs are challenging to implement and it may happen we will have to rely on the coverage pool (https://github.com/keep-network/coverage-pools) and DAO to reimburse unlucky depositors and bring back the balance to the system in case of a wallet fraud by liquidating a part of the coverage pool manually. 

The probability of 51-of-100 wallet being fraudulent is negligible: https://github.com/keep-network/tbtc-v2/blob/main/docs/rfc/rfc-2.adoc#111-group-size-and-threshold and the coverage pool would be there to bring the balance back in case we are unlucky and malicious wallet emerges.

We do not want to slash for a misbehavior that is not provable on-chain and it is possible to construct such a Bitcoin transaction that is not provable on Ethereum, see https://consensys.net/diligence/blog/2020/05/tbtc-navigating-the-cross-chain-conundrum/

The algorithm described below assumes we will be able to prove the TX on Ethereum which may not always be the case. Consider the steps below as an idea, and not necessarily how this function will be implemented because it may happen this function will never be implemented, given the Bitcoin transaction size problems.